ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/shared/target/TargetLocationsGroup.java

```diff
@@ -563,6 +578,35 @@ public IStatus runInUIThread(IProgressMonitor monitor) {
 
 	}
 
+	private void handleClean() {
+		ITargetPlatformService service = PDECore.getDefault().acquireService(ITargetPlatformService.class);
+		if (service != null && fTarget != null) {
+			try {
+				ITargetHandle currentTarget = service.getWorkspaceTargetHandle();
+				if (fTarget.getHandle().equals(currentTarget)) {
+					getOrCreateCleanJob().schedule();
+				}
+			} catch (CoreException e) {
+
+			}
+		}
+	}
+
+	private Job getOrCreateCleanJob() {
+		if (cleanJob == null) {
+			cleanJob = new Job(PDEUIMessages.TargetPlatformPreferencePage2_26) {
```

Comment:
Using the `Job.create()` factory should make the code cleaner.

Furthermore you probably want to hold the `workspace` rule to prevent multiple of those jobs from running at the same time.
I assume you tried to prevent that by saving the job as field, don't you?
That is probably not the savest way. It is probably possible to create multiple instances of the surrounding class if one quickly closes and reopens the editor. Then the job of the previous editor instance would still be running.
You could also add a job family and search for existing jobs.

Furthermore just rescheduling the job on each button click is IMHO not ideal. We could consider to disable the button when the job starts and re-enable once the GC has completed.

Reply:
I will make some improvements as suggested. See one of my comments: This job has always run so quickly that I haven't even been able to read the progress view's text before it disappears.